src/metabase/query_processor/util/add_alias_info.clj

```diff
@@ -39,552 +39,413 @@
+                                  (update-in (into [::original] join-path)
+                                             (fn [join]
+                                               (update join :stages (fn [stages]
+                                                                      (conj (vec stages) {:lib/type :mbql.stage/mbql}))))))
+        ;; e.g. if join has one stage, `join-last-stage-path` could be something like `[:stages 0 :joins 0 :stages 0]`
+        ;; and `join-extra-stage-path` would be `[:stages 0 :joins 0 :stages 1]`
+        join-last-stage-path  (lib.walk/join-last-stage-path join-path join)
+        join-extra-stage-path (-> (vec join-last-stage-path)
+                                  pop
+                                  (conj (inc (last join-last-stage-path))))
+        col                   (resolve-field-ref
+                               query'
+                               join-extra-stage-path
+                               ;; join alias is for use outside the join, a column doesn't have the join alias INSIDE
+                               ;; the join itself.
+                               (lib/with-join-alias field-ref nil))
+        last-stage-alias      (or (:lib/source-column-alias col)
+                                  (throw (ex-info "Expected resolved column to have :lib/source-column-alias"
+                                                  {:path join-path
+                                                   :join join
+                                                   :ref  field-ref
+                                                   :col  col})))
+        source-alias          (escaped-desired-alias query (lib.walk/join-last-stage-path join-path join) last-stage-alias)
+        source-table          (escaped-join-alias query parent-stage-path (:join-alias opts))]
+    ;; don't need to calculate `::desired-alias` because it may not be returned and even if it is it's not getting
+    ;; returned in the join conditions
+    (lib/update-options field-ref assoc
+                        ::source-alias source-alias
+                        ::source-table source-table)))
+
+(mu/defn- add-alias-info-to-join-conditions :- ::lib.schema.join/join
+  "Add alias info to refs in join `:conditions`. Join `:fields` do not need to be updated since they are only used to
+  update parent stage `:fields` as appropriate (which will have already been done by now) and otherwise do not directly
+  affect the resulting SQL (i.e., drivers should not be looking at them anyway)."
+  [query     :- ::lib.schema/query
+   join-path :- ::lib.walk/path
+   join      :- ::lib.schema.join/join]
+  (let [parent-stage-path (lib.walk/join-parent-stage-path join-path)
+        update-other-ref  (fn [field-ref]
+                            (let [col (resolve-field-ref query parent-stage-path field-ref)]
+                              (update-field-ref query parent-stage-path field-ref col)))
```

Comment:
`update-field-ref` is a pretty vague name. Looking at what it does, though, I'm not sure of a better name.